File: apps/agent/src/context/calendarDensity.ts
/**
 * Pre-fetched calendar density (today + tomorrow event counts) injected into
 * the system prompt every turn. The LLM gets narrative awareness of "you have
 * 8 meetings tomorrow" without having to call call_workspace — saves a turn
 * per chat for the common "anything on today?" / "busy tomorrow?" prompts.
 *
 * Auth: reuses workspaceTokensStore.getValidAccessToken(uid). Token never
 * appears in argv — passed via env to the gws subprocess (same pattern as
 * call_workspace).
 *
 * Caching: per (uid, tz) for 5 min. Calendars shift fast and the data is
 * cheap to refresh, but caching across rapid successive turns is the win.
 *
 * Failure mode: any error → return null. The format helper omits the block
 * when null. Never blocks a turn.
 */

import { ScopeRequiredError, type WorkspaceTokensStore } from '../storage/workspaceTokens.js';
import { type ExecFileLike, defaultExecFile } from '../tools/callWorkspace.js';

export interface CalendarDensitySummary {
  today: {
    count: number;
    /** HH:MM in user's tz; null when no events with explicit start times. */
    firstStart: string | null;
    /** HH:MM in user's tz; null when no events with explicit end times. */
    lastEnd: string | null;
    /** HH:MM of next not-yet-started event today, or null if all are past. */
    nextStart: string | null;
  };
  tomorrow: {
    count: number;
    firstStart: string | null;
    lastEnd: string | null;
  };
}

export interface CalendarDensityClient {
  get(params: {
    uid: string;
    timezone: string;
    now: Date;
  }): Promise<CalendarDensitySummary | null>;
}

interface CacheEntry {
  at: number;
  /** The tz-local "today" date this entry was computed for — invalidates on day rollover. */
  todayDate: string;
  value: CalendarDensitySummary | null;
}

const TTL_MS = 5 * 60_000;
const TIMEOUT_MS = 10_000;

interface GoogleEventTime {
  dateTime?: string;
  date?: string;
}

interface GoogleEvent {
  start?: GoogleEventTime;
  end?: GoogleEventTime;
}

interface GwsListResponse {
  items?: GoogleEvent[];
}

export interface CreateCalendarDensityClientDeps {
  store: WorkspaceTokensStore;
  exec?: ExecFileLike;
  ttlMs?: number;
  gwsPath?: string;
  timeoutMs?: number;
}

export function createCalendarDensityClient(
  deps: CreateCalendarDensityClientDeps,
): CalendarDensityClient {
  const { store } = deps;
  const exec = deps.exec ?? defaultExecFile;
  const ttl = deps.ttlMs ?? TTL_MS;
  const gwsPath = deps.gwsPath ?? 'gws';
  const timeoutMs = deps.timeoutMs ?? TIMEOUT_MS;
  const cache = new Map<string, CacheEntry>();

  return {
    async get({ uid, timezone, now }) {
      const todayDate = dateInTz(now, timezone);
      const key = `${uid}:${timezone}`;
      const hit = cache.get(key);
      if (hit && hit.todayDate === todayDate && Date.now() - hit.at < ttl) {
        return hit.value;
      }

      let token: string;
      try {
        token = await store.getValidAccessToken(uid);
      } catch (err) {
        if (err instanceof ScopeRequiredError) {
          // No valid token. Don't cache — state machine will drop us out of
          // workspace_connected next turn and we'll stop being called.
          return null;
        }
        return null;
      }

      // Wide UTC window comfortably covers today + tomorrow in any timezone
      // (max ±14h offset). Keeps us off the timezone-arithmetic hot path.
      const timeMin = new Date(now.getTime() - 24 * 60 * 60_000).toISOString();
      const timeMax = new Date(now.getTime() + 48 * 60 * 60_000).toISOString();
      const params = JSON.stringify({
        calendarId: 'primary',
        timeMin,
        timeMax,
        singleEvents: true,
        orderBy: 'startTime',
        maxResults: 50,
      });

      let res: { stdout: string; stderr: string; code: number | null };
      try {
        res = await exec(gwsPath, ['calendar', 'events', 'list', '--params', params], {
          env: { ...process.env, GOOGLE_WORKSPACE_CLI_TOKEN: token },
          timeout: timeoutMs,
        });
      } catch {
        return null;
      }
      if (res.code !== 0) return null;

      let parsed: GwsListResponse;
      try {
        parsed = JSON.parse(res.stdout) as GwsListResponse;
      } catch {
        return null;
      }
      const items = parsed.items ?? [];

      const tomorrowDate = dateInTz(new Date(now.getTime() + 24 * 60 * 60_000), timezone);
      const summary = bucket(items, todayDate, tomorrowDate, timezone, now);
      cache.set(key, { at: Date.now(), todayDate, value: summary });
      return summary;
    },
  };
}

function bucket(
  items: GoogleEvent[],
  todayDate: string,
  tomorrowDate: string,
  tz: string,
  now: Date,
): CalendarDensitySummary {
  const todayStarts: Date[] = [];
  const todayEnds: Date[] = [];
  let todayCount = 0;
  const tomorrowStarts: Date[] = [];
  const tomorrowEnds: Date[] = [];
  let tomorrowCount = 0;

  for (const item of items) {
    const startStr = item.start?.dateTime ?? item.start?.date;
    if (!startStr) continue;
    // For all-day events, the date string is YYYY-MM-DD with no time. We
    // count them but don't fold them into the start/end time fields.
    const isAllDay = !item.start?.dateTime;
    const eventDay = isAllDay ? (startStr as string) : dateInTz(new Date(startStr), tz);
    if (eventDay === todayDate) {
      todayCount += 1;
      if (!isAllDay && item.start?.dateTime) todayStarts.push(new Date(item.start.dateTime));
      if (!isAllDay && item.end?.dateTime) todayEnds.push(new Date(item.end.dateTime));
    } else if (eventDay === tomorrowDate) {
      tomorrowCount += 1;
      if (!isAllDay && item.start?.dateTime) tomorrowStarts.push(new Date(item.start.dateTime));
      if (!isAllDay && item.end?.dateTime) tomorrowEnds.push(new Date(item.end.dateTime));
    }
  }

  const firstTodayStart = minDate(todayStarts);
  const lastTodayEnd = maxDate(todayEnds);
  const nextTodayStart = todayStarts
    .filter((d) => d.getTime() >= now.getTime())
    .sort((a, b) => a.getTime() - b.getTime())[0];

  return {
    today: {
      count: todayCount,
      firstStart: firstTodayStart ? timeInTz(firstTodayStart, tz) : null,
      lastEnd: lastTodayEnd ? timeInTz(lastTodayEnd, tz) : null,
      nextStart: nextTodayStart ? timeInTz(nextTodayStart, tz) : null,
    },
    tomorrow: {
      count: tomorrowCount,
      firstStart: minDate(tomorrowStarts) ? timeInTz(minDate(tomorrowStarts) as Date, tz) : null,
      lastEnd: maxDate(tomorrowEnds) ? timeInTz(maxDate(tomorrowEnds) as Date, tz) : null,
    },
  };
}

function minDate(ds: Date[]): Date | undefined {
  if (ds.length === 0) return undefined;
  return ds.reduce((a, b) => (a.getTime() <= b.getTime() ? a : b));
}

function maxDate(ds: Date[]): Date | undefined {
  if (ds.length === 0) return undefined;
  return ds.reduce((a, b) => (a.getTime() >= b.getTime() ? a : b));
}

function dateInTz(d: Date, tz: string): string {
  // en-CA gives YYYY-MM-DD reliably across all timezones.
  return new Intl.DateTimeFormat('en-CA', {
    timeZone: tz,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(d);
}

function timeInTz(d: Date, tz: string): string {
  // sv-SE returns HH:MM in 24-hour format; en-GB has historical edge cases
  // around midnight (24:00 vs 00:00) — sv-SE doesn't.
  return new Intl.DateTimeFormat('sv-SE', {
    timeZone: tz,
    hour: '2-digit',
    minute: '2-digit',
  }).format(d);
}

File: apps/agent_py/src/lifecoach_agent/context/calendar_density.py
"""Pre-fetched calendar density (today + tomorrow event counts) injected
into the system prompt every turn.

Mirrors `apps/agent/src/context/calendarDensity.ts` minus the
`gws` CLI subprocess. Phase 7 plugs a `google-api-python-client`-backed
fetcher into `EventsFetcher`; until then this module is callable but
fetcher implementations are stubs in tests.

Auth: defers to a `WorkspaceTokensProtocol` (Phase 5 implements it).
Caching: per `(uid, tz)`, 5 minutes, invalidated on day rollover.
Failure mode: any error → return None. Never blocks a turn.
"""

from __future__ import annotations

import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any, Protocol
from zoneinfo import ZoneInfo

from lifecoach_agent.prompt.build_instruction import (
    CalendarDay,
    CalendarDensitySummary,
    TodayEvent,
)

TODAY_EVENT_LIMIT = 10
TTL_S = 5 * 60.0


class WorkspaceTokensProtocol(Protocol):
    """The workspace-tokens store. Phase 5 implements this against the
    Firestore-backed token doc."""

    async def get_valid_access_token(self, uid: str) -> str: ...


class ScopeRequiredError(Exception):
    """Raised by `WorkspaceTokensProtocol.get_valid_access_token` when
    the user has lapsed and needs to reconnect."""


# Function fetching events for `uid` between `time_min` and `time_max`
# (RFC3339 strings). Returns a list of dict-shaped events as the Calendar
# API returns them; we only read `summary`, `start.dateTime|date`,
# `end.dateTime|date`. Phase 7 implements this with `google-api-python-client`.
EventsFetcher = Callable[[str, str, str, str], Awaitable[list[dict[str, Any]] | None]]


@dataclass
class _CacheEntry:
    at: float
    today_date: str
    value: CalendarDensitySummary | None


def _date_in_tz(d: datetime, tz: str) -> str:
    return d.astimezone(ZoneInfo(tz)).strftime("%Y-%m-%d")


def _time_in_tz(d: datetime, tz: str) -> str:
    return d.astimezone(ZoneInfo(tz)).strftime("%H:%M")


def _parse_iso(s: str) -> datetime:
    """Parse an RFC3339 string into a tz-aware datetime."""
    return datetime.fromisoformat(s.replace("Z", "+00:00"))


class CalendarDensityClient:
    def __init__(
        self,
        *,
        store: WorkspaceTokensProtocol,
        events_fetcher: EventsFetcher,
        ttl_s: float = TTL_S,
        now: Callable[[], float] | None = None,
    ) -> None:
        self._store = store
        self._fetch = events_fetcher
        self._ttl = ttl_s
        self._now = now or time.monotonic
        self._cache: dict[str, _CacheEntry] = {}

    async def get(self, *, uid: str, timezone: str, now: datetime) -> CalendarDensitySummary | None:
        if now.tzinfo is None:
            now = now.replace(tzinfo=UTC)
        today_date = _date_in_tz(now, timezone)
        key = f"{uid}:{timezone}"
        now_t = self._now()
        hit = self._cache.get(key)
        if hit and hit.today_date == today_date and now_t - hit.at < self._ttl:
            return hit.value

        try:
            access_token = await self._store.get_valid_access_token(uid)
        except ScopeRequiredError:
            return None
        except Exception:  # noqa: BLE001
            return None

        time_min = (now - timedelta(hours=24)).astimezone(UTC).isoformat().replace("+00:00", "Z")
        time_max = (now + timedelta(hours=48)).astimezone(UTC).isoformat().replace("+00:00", "Z")
        try:
            items = await self._fetch(access_token, "primary", time_min, time_max)
        except Exception:  # noqa: BLE001
            return None
        if items is None:
            return None

        tomorrow_date = _date_in_tz(now + timedelta(days=1), timezone)
        summary = _bucket(items, today_date, tomorrow_date, timezone, now)
        self._cache[key] = _CacheEntry(at=now_t, today_date=today_date, value=summary)
        return summary


def _bucket(
    items: list[dict[str, Any]],
    today_date: str,
    tomorrow_date: str,
    tz: str,
    now: datetime,
) -> CalendarDensitySummary:
    today_starts: list[datetime] = []
    today_ends: list[datetime] = []
    today_events: list[tuple[float, TodayEvent]] = []
    today_count = 0
    tomorrow_starts: list[datetime] = []
    tomorrow_ends: list[datetime] = []
    tomorrow_count = 0

    for item in items:
        start = item.get("start") or {}
        start_str = start.get("dateTime") or start.get("date")
        if not start_str:
            continue
        is_all_day = "dateTime" not in start
        event_day = start_str if is_all_day else _date_in_tz(_parse_iso(start_str), tz)
        end = item.get("end") or {}
        if event_day == today_date:
            today_count += 1
            start_dt = _parse_iso(start["dateTime"]) if "dateTime" in start else None
            end_dt = _parse_iso(end["dateTime"]) if "dateTime" in end else None
            if not is_all_day and start_dt:
                today_starts.append(start_dt)
            if not is_all_day and end_dt:
                today_ends.append(end_dt)
            today_events.append(
                (
                    start_dt.timestamp() if start_dt else 0.0,
                    TodayEvent(
                        summary=(item.get("summary") or "").strip() or "(no title)",
                        start=_time_in_tz(start_dt, tz) if start_dt else None,
                        end=_time_in_tz(end_dt, tz) if end_dt else None,
                        allDay=is_all_day,
                    ),
                )
            )
        elif event_day == tomorrow_date:
            tomorrow_count += 1
            if not is_all_day and "dateTime" in start:
                tomorrow_starts.append(_parse_iso(start["dateTime"]))
            if not is_all_day and "dateTime" in end:
                tomorrow_ends.append(_parse_iso(end["dateTime"]))

    first_today = min(today_starts) if today_starts else None
    last_today = max(today_ends) if today_ends else None
    upcoming_today = sorted([d for d in today_starts if d >= now], key=lambda d: d.timestamp())
    next_today = upcoming_today[0] if upcoming_today else None

    today_events_sorted = [e for _k, e in sorted(today_events, key=lambda kv: kv[0])][
        :TODAY_EVENT_LIMIT
    ]

    return CalendarDensitySummary(
        today=CalendarDay(
            count=today_count,
            firstStart=_time_in_tz(first_today, tz) if first_today else None,
            lastEnd=_time_in_tz(last_today, tz) if last_today else None,
            nextStart=_time_in_tz(next_today, tz) if next_today else None,
            events=today_events_sorted,
        ),
        tomorrow=CalendarDay(
            count=tomorrow_count,
            firstStart=_time_in_tz(min(tomorrow_starts), tz) if tomorrow_starts else None,
            lastEnd=_time_in_tz(max(tomorrow_ends), tz) if tomorrow_ends else None,
            nextStart=None,
            events=[],
        ),
    )
